src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs

```diff
@@ -0,0 +1,301 @@
+        {
+            if (query.Length == 0)
+                return;
+
+            var decoded = HtmlDecode(query);
+            var decodedLength = decoded.Length;
+            var namePos = 0;
+            var first = true;
+            while (namePos <= decodedLength)
+            {
+                int valuePos = -1, valueEnd = -1;
+                for (var q = namePos; q < decodedLength; q++)
+                    if ((valuePos == -1) && (decoded[q] == '='))
+                    {
+                        valuePos = q + 1;
+                    }
+                    else if (decoded[q] == '&')
+                    {
+                        valueEnd = q;
+                        break;
+                    }
+
+                if (first)
+                {
+                    first = false;
+                    if (decoded[namePos] == '?')
+                        namePos++;
+                }
+
+                string name;
+                if (valuePos == -1)
+                {
+                    name = null;
+                    valuePos = namePos;
+                }
+                else
+                {
+                    name = UrlDecode(decoded.Substring(namePos, valuePos - namePos - 1), encoding);
+                }
+                if (valueEnd < 0)
+                {
+                    namePos = -1;
+                    valueEnd = decoded.Length;
+                }
+                else
+                {
+                    namePos = valueEnd + 1;
+                }
+                var value = UrlDecode(decoded.Substring(valuePos, valueEnd - valuePos), encoding);
+
+                result.Add(name, value);
+                if (namePos == -1)
+                    break;
+            }
+        }
+
+        public static string HtmlDecode(string s)
+        {
+            return HttpEncoder.HtmlDecode(s);
+        }
+
+
+        public static void HtmlDecode(string s, TextWriter output)
+        {
+            HttpEncoder.HtmlDecode(s, output);
+        }
+
+
+        public static string HtmlEncode(string s)
+        {
+            return HttpEncoder.HtmlEncode(s);
+        }
+
+
+        public static string HtmlEncode(object value)
+        {
+            if (value == null)
+                return null;
+
+            return HtmlEncode(Convert.ToString(value, CultureInfo.CurrentCulture));
+        }
+
+
+        public static void HtmlEncode(string s, TextWriter output)
+        {
+            HttpEncoder.HtmlEncode(s, output);
+        }
+
+
+        public static string HtmlAttributeEncode(string s)
+        {
+            return HttpEncoder.HtmlAttributeEncode(s);
+        }
+
+
+        public static void HtmlAttributeEncode(string s, TextWriter output)
+        {
+            HttpEncoder.HtmlAttributeEncode(s, output);
+        }
+
+        public static string UrlEncode(string str)
+        {
+            if (str == null)
+                return null;
+            return UrlEncode(str, Encoding.UTF8);
```

Comment:
That's really unfortunate.  I agree with Wes; we should find a way to share the code.  Can you please open an issue to track that?  The existing code has had a fair number of optimizations applied to it since it came to the repo.  We could, for example, augment that code with support for encoding (even though it won't be used via those public APIs), and compile the code into both assemblies.  Or add public API there that supports encoding and simply call into it.